newhelm/record_init.py

```diff
@@ -0,0 +1,46 @@
+import importlib
+from typing import Any, List, Mapping
+from pydantic import BaseModel
+
+
+class InitializationRecord(BaseModel):
+    """Holds data sufficient to reconstruct an object."""
+
+    module: str
+    qual_name: str
+    args: List[Any]
+    kwargs: Mapping[str, Any]
+
+    def recreate_object(self):
+        """Redoes the init call from this record."""
+        cls = getattr(importlib.import_module(self.module), self.qual_name)
+        return cls(*self.args, **self.kwargs)
+
+
+def record_init(init):
+    """Decorator for the __init__ function to store what arguments were passed."""
+
+    def wrapped_init(*args, **kwargs):
```

Comment:
Are we not `functools.wrap` enjoyers? - Please ignore if there is some reason we aren't making use of this for this particular implementation.

Reply:
I'd forgotten that utility existed. Added.